TAG initial implementation

jakule opened the pull request:
This change introduces the initial Teleport integration with TAG. The implementation introduces:

* new `access_graph` section in `teleport.yaml` file that holds the access graph configuration
* new `access_graph` RBAC kind, that controls who has access to TAG feature
* the new access graph RBAC kind is added to the ACL REST endpoint, so WebUI know if TAG is enabled and the user has access to it.

changelog: Add Teleport Access Graph integration

jakule committed:
TAG initial implementation

This change introduces the initial Teleport integration with TAG. The implementation introduces:

* new `access_graph` section in `teleport.yaml` file that holds the access graph configuration
* new `access_graph` RBAC kind, that controls who has access to TAG feature
* the new access graph RBAC kind is added to the ACL REST endpoint, so WebUI know if TAG is enabled and the user has access to it.

r0mant commented on lib/auth/grpcserver.go:
```diff
@@ -389,8 +390,22 @@ func (g *GRPCServer) WatchEvents(watch *authpb.Watch, stream authpb.AuthService_
+	return trace.Wrap(WatchEvents(watch, stream, auth.User.GetName(), auth))
+}
+
+type WatchEvent interface {
+	Context() context.Context
+	Send(*authpb.Event) error
+}
+
+type Watcher interface {
+	NewStream(ctx context.Context, watch types.Watch) (stream.Stream[types.Event], error)
+}
+
+func WatchEvents(watch *authpb.Watch, stream WatchEvent, componentName string, auth Watcher) error {
```
Nit: godocs.

r0mant commented on lib/config/fileconf.go:
```diff
@@ -96,6 +96,9 @@ type FileConfig struct {
 	Plugins PluginService `yaml:"plugin_service,omitempty"`
+
+	// AccessGraph is the section of the config describing AccessGraph service
+	AccessGraph AccessGraph `yaml:"access_graph,omitempty"`
```
Nit: I don't think "omitempty" does anything here tbh but I see other fields have it also so 🤷

jakule replied:
Even if it's pointless at least it's consistent :sweat_smile:

r0mant commented on lib/config/fileconf.go:
```diff
@@ -815,6 +818,15 @@ type PluginService struct {
 
+// AccessGraph represents the configuration for the AccessGraph service.
+type AccessGraph struct {
+	Enabled  bool   `yaml:"enabled"`
+	UseAuth  bool   `yaml:"use_auth"`
+	Endpoint string `yaml:"endpoint"`
+	CA       string `yaml:"ca"`
+	Insecure bool   `yaml:"insecure"`
```
Godocs please.

r0mant commented on lib/service/servicecfg/config.go:
```diff
@@ -308,6 +311,24 @@ func WithKubeMultiplexerIgnoreSelfConnectionsOption() KubeMultiplexerIgnoreSelfC
 
+// AccessGraphConfig represents TAG server config
+type AccessGraphConfig struct {
+	// Enabled Access Graph reporting enabled
+	Enabled bool
+
+	// When enabled, the TAG server will be used to evaluate access requests
+	UseAuth bool
```
This is a bit confusing tbh, so "Enabled" just turns on syncing of users/roles/resources to TAG, but "UseAuth" turns on actual RBAC checks? Would it make sense to rename these to something like `SyncEnabled bool` and `RBACEnabled bool` to make it more clear?

jakule replied:
My bad. "UseAuth" should be removed, as we're not shipping this functionality.

jakule committed:
Remove `UseAuth` and improve comments in `AccessGraph`

This commit removes the `UseAuth` field from the `AccessGraph` configuration. At the same time, the comments for each field in `AccessGraph` have been made more consistent and descriptive, providing clear declaration of each field's purpose. The comments in `WatchEvents` in `grpcserver.go` have also been updated for clarity.
